R/evaluate-clusters.R

```diff
@@ -307,3 +307,125 @@ calculate_stability <- function(
+#' object provided.
+#'
+#' @param x An object containing PCs that clusters were calculated from. This can be
+#'   either a SingleCellExperiment object, a Seurat object, or a matrix where columns
+#'   are PCs and rows are cells. If a matrix is provided, it must have row names of cell
+#'   ids (e.g., barcodes).
+#' @param cluster_results A single data frame or list of data frames obtained from
+#'   `rOpenScPCA::sweep_clusters()`. Each data frame in the list should contains
+#'   at least two columns: one representing unique cell ids, and one containing
+#'   cluster assignments. By default, these columns should be named `cell_id` and
+#'   `cluster` respectively, though this can be customized. The cell id column's
+#'   values should match either the PC matrix row names, or the
+#'   SingleCellExperiment/Seurat object cell ids. Typically this data frame will be
+#'   output from the `rOpenScPCA::calculate_clusters()` function.
+#' @param metrics Which metrics should be collected? Options are one or both of "purity" or "silhouette".
+#' Default is to collect both purity and silhouette.
+#'
+#' @return An updated list of data frames with additional columns from running evaluation with
+#' `rOpenScPCA::calculate_silhouette()` and/or `rOpenScPCA::calculate_purity()`, based on
+#'   `calculate_silhouette()` functions output.
+#'
+#' @export
+#'
+#' @examples
+#' \dontrun{
+#'
+#' # Setting the seed is a good idea
+#' set.seed(2024)
+#'
+#' # Calculate Principal Components
+#' pca_matrix <- reducedDim(sce_object, "PCA")
+#'
+#' # We can put in a single data frame of cluster results:
+#' cluster_df <- calculate_clusters(
+#'   pca_matrix,
+#'   algorithm = "leiden",
+#'   resolution = 0.1,
+#'   seed = 11
+#' )
+#'
+#' # Then we can evaluate these cluster stats with
+#' # calculate_cell_cluster_metrics:
+#' sweep_list_evaled <- calculate_cell_cluster_metrics(
+#'   x = pca_matrix,
+#'   cluster_results = cluster_df
+#' )
+#'
+#' ######### Evaluate a list of multiple cluster results ############
+#' # If we obtain a list of clusters like so...
+#' sweep_list <- sweep_clusters(
+#'   sce_object,
+#'   algorithm = "walktrap",
+#'   weighting = "jaccard",
+#'   nn = c(10, 15, 25),
+#'   resolution = c(0.75, 1),
+#'   seed = 9
+#' )
+#'
+#' # Then we can evaluate these cluster stats with
+#' # calculate_cell_cluster_metrics:
+#' sweep_list_evaled <- calculate_cell_cluster_metrics(
+#'   x = pc_mat,
+#'   cluster_results = sweep_list
+#' )
+#' }
+#'
+calculate_cell_cluster_metrics <- function(x,
+                                           cluster_results,
+                                           metrics = c("purity", "silhouette")) {
+  supported_evals <- c("purity", "silhouette")
+
+  if (is.data.frame(cluster_results)) {
+    cluster_results <- list(cluster_results)
+  }
+  # Check input arguments
+  stopifnot(
+    "`cluster_results` must be a list containing data.frames" =
+      is.list(cluster_results) && is.data.frame(cluster_results[[1]]),
```

Comment:
We can check them all like this:

```suggestion
      is.list(cluster_results) &&  all(sapply(cluster_results, is.data.frame)),
```